stdenv: log hooks as they run (take II)

Qyriad opened the pull request:
## Description of changes

A second take at #290081, which changes the stdenv builder to log setup and build hooks as they run, but which was reverted for the extra logging during the internals of `nix-shell -p`. This PR does the same logging, but to $NIX_LOG_FD instead, which is echoed during any normal build, but not during the internals of `nix-shell -p`.

cc @philiptaron @RaitoBezarius from the last PR

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Qyriad commented:
Rebased on latest staging

RaitoBezarius commented:
I would recommend to wait for the next staging cycle as the current one is the last NixOS 24.05's one.

philiptaron commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -47,9 +47,59 @@ getAllOutputNames() {
 ######################################################################
 # Hook handling.
 
+# Log a hook, to be run before the hook is actually called.
+# logging for "implicit" hooks -- the ones specified directly
+# in derivation's arguments -- is done in _callImplicitHook instead.
+_logHook() {
+    local hookKind="$1"
+    local hookExpr="$2"
+    shift 2
+
+    if declare -F "$hookExpr" > /dev/null 2>&1; then
+        nixLog "calling '$hookKind' function hook '$hookExpr'" "$@"
+    elif type -p "$hookExpr" > /dev/null; then
+        nixLog "sourcing '$hookKind' script hook '$hookExpr'"
+    elif [[ "$hookExpr" != "_callImplicitHook"* ]]; then
+        # Here we have a string hook to eval.
+        # Join lines onto one with literal \n characters unless NIX_DEBUG >= 2.
+        local exprToOutput
+        if (( "${NIX_DEBUG:-0}" >= 2)); then
+            exprToOutput="$hookExpr"
+        else
+            while IFS= read -r hookExprLine; do
+                # These lines often have indentation,
+                # so let's remove leading whitespace.
+                hookExprLine="${hookExprLine#"${hookExprLine%%[![:space:]]*}"}"
+                # If this line wasn't entirely whitespace,
+                # then add it to our output
+                if [[ -n "$hookExprLine" ]]; then
+                    exprToOutput+="$hookExprLine\\n "
+                fi
+            done <<< "$hookExpr"
+
+            # And then remove the final, unnecessary, \n
+            exprToOutput="${exprToOutput%%\\n }"
+        fi
+        nixLog "evaling '$hookKind' string hook '$exprToOutput'"
+    fi
+}
```
Since this function only calls through to `nixLog`, which is a no-op when `NIX_LOG_FD` does not exist, consider putting it in the same if branch.

Qyriad replied:
Good idea! No need to perform all that work for nothing

philiptaron commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -1558,9 +1615,7 @@ runPhase() {
     if [[ "$curPhase" = installCheckPhase && -z "${doInstallCheck:-}" ]]; then return; fi
     if [[ "$curPhase" = distPhase && -z "${doDist:-}" ]]; then return; fi
 
-    if [[ -n $NIX_LOG_FD ]]; then
-        echo "@nix { \"action\": \"setPhase\", \"phase\": \"$curPhase\" }" >&"$NIX_LOG_FD"
-    fi
```
Glad to see this change!

philiptaron reviewed (approved):
Code looks good to me. I'm currently building a representative sample of nixpkgs, and I see the hook logging.

xzfc commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -47,10 +47,65 @@ getAllOutputNames() {
+            if (( "${NIX_DEBUG:-0}" >= 2)); then
+                exprToOutput="$hookExpr"
+            else
+                while IFS= read -r hookExprLine; do
```
Nit: add `local hookExprLine`.

xzfc commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -47,10 +47,65 @@ getAllOutputNames() {
+            # Here we have a string hook to eval.
+            # Join lines onto one with literal \n characters unless NIX_DEBUG >= 2.
+            local exprToOutput
+            if (( "${NIX_DEBUG:-0}" >= 2)); then
```
```suggestion
            if (( "${NIX_DEBUG:-0}" >= 2 )); then
```

Qyriad committed:
stdenv: log hooks as they run (take II)

A second take at eb28e5e72ef9, which was reverted for the extra logging
during the internals of `nix-shell -p`. This commit does the same
logging, but to $NIX_LOG_FD instead, which is echoed during any normal
build, but not during the internals of `nix-shell -p`.

[1]: eb28e5e72ef912629ded3e265f7344ca21d115b9

paparodeo reviewed (approved):
<details><summary>LGTM tested with nix-shell on master</summary>

```console
$ curl -L https://github.com/NixOS/nixpkgs/pull/310387.patch | git apply
$ cat test.sh
#!/usr/bin/env nix-shell
#! nix-shell -i bash --pure
#! nix-shell -p hello
hello
$ NIX_PATH=nixpkgs=`pwd` ./test.sh  | tee stdout.log
[...]
$ cat stdout.log
Hello, world!
```
</details>

philiptaron commented:
Let's rumble! Thanks to everyone for getting this through and thanks to @qyriad for authoring it. I really like how it helps make what's going on with hooks more understandable.
